python/python_fastapi_server/routers/caching_middleware.py

```diff
@@ -0,0 +1,147 @@
+from io import BytesIO
+import os
+from urllib.parse import urlsplit
+
+from starlette.middleware.base import BaseHTTPMiddleware
+from starlette.responses import Response
+from fastapi import BackgroundTasks
+
+import calendar
+from datetime import datetime
+import redis.asyncio as redis  # This can also be used to connect to a Redis cluster
+
+# from redis.asyncio.cluster import RedisCluster as Redis  # Cluster client, for testing
+
+import json
+import brotli
+
+ADAGUC_REDIS = os.environ.get("ADAGUC_REDIS")
+
+MAX_SIZE_FOR_CACHING = 10000000
+
+
+async def get_cached_response(redis_pool, request):
+    key = generate_key(request)
+    redis_client = redis.Redis(connection_pool=redis_pool)
+    cached = await redis_client.get(key)
+    await redis_client.aclose()
+    if not cached:
+        return None, None, None
+
+    entrytime = int(cached[:10].decode("utf-8"))
+    currenttime = calendar.timegm(datetime.utcnow().utctimetuple())
+    age = currenttime - entrytime
+
+    headers_len = int(cached[10:16].decode("utf-8"))
+    headers = json.loads(cached[16 : 16 + headers_len].decode("utf-8"))
+
+    data = brotli.decompress(cached[16 + headers_len :])
+    return age, headers, data
+
+
+headers_to_skip = ["x-process-time", "age"]
+
+
+async def response_to_cache(redis_pool, request, headers, data, ex: int):
+    key = generate_key(request)
+
+    fixed_headers = {}
+    for k in headers:
+        if not k in headers_to_skip:
+            fixed_headers[k] = headers[k]
+    headers_json = json.dumps(fixed_headers, ensure_ascii=False).encode("utf-8")
+
+    entrytime = f"{calendar.timegm(datetime.utcnow().utctimetuple()):10d}".encode(
+        "utf-8"
+    )
+    compressed_data = brotli.compress(data)
+    if len(compressed_data) < MAX_SIZE_FOR_CACHING:
+        redis_client = redis.Redis(connection_pool=redis_pool)
+        await redis_client.set(
+            key,
+            entrytime
+            + f"{len(headers_json):06d}".encode("utf-8")
+            + headers_json
+            + brotli.compress(data),
+            ex=ex,
+        )
+        await redis_client.aclose()
+
+
+def generate_key(request):
+    key = request.url.path.encode("utf-8")
+    if len(request["query_string"]) > 0:
+        key += b"?" + request["query_string"]
```

Comment:
Ik snap niet hoe dit werkt. `b"?"`, en `request["query_string"]` is toch een string? In mijn test mag ik die niet optellen. Maar toch lijkt het te werken. Of is `request["query_string"]` al binary?

Reply:
Dit is een manier om byte arrays te concateneren

Reply:
Ok, dus `request["query_string"]` is al binary.